Create 'sustainable coding and reuseability' principle (#192)
HO-CTO/engineering-guidance-and-standards: docs/principles/sustainable-code.md

---
layout: sub-navigation
order: 1
title: Write sustainable and evolutionary code
date: git Last Modified
tags:
  - Ways of working
  - Software design
  - Sustainable coding
  - Reusability
---

{% import "macros/tags.njk" as pageTag %}

Last updated: {{ page.date | postDate }}
{{ pageTag.tags(tags)  }}

When writing code, keep in mind that this should be clean, reusable and descriptive. Reuse existing artefacts before delivering new ones where possible. This can enable code to be easier to understand and cheaper to modify and reuse.

---

## Rationale

> This aides in allowing your code to be further developed and evolved in the future, and reduces development time and costs where code is reused. 

Reviewer: ```suggestion
Writing code takes time and effort. Code is read more than it is modified, and modified more that it is written. It is worth spending more time and effort to write code that it is easier to understand and to extend with new functionality. 

Reusing existing code saves considerable development time and effort at the cost of additional complexity. Writing code that is understandable and modular minimises this complexity so that code can be more effectivly reused. Writing code that is easy to modify allows this complexity to only be added when it is useful.
```

---

## Applications and Implications

> - Keep your code as simple and understandable as possible. Start with only the code needed for the task at hand. Use refactorings where needed to separate concerns, facilitate reuse, and improve readability  when the need arises.

Reviewer: ```suggestion
- Keep your code as simple and understandable as possible. Start with only the code needed for the task at hand. Use refactorings where needed to separate concerns, facilitate reuse, and improve readability when the need arises.
```

---
layout: sub-navigation
order: 1

> title: Write sustainable and evolutionary code

Author: Title change sounds good to me. I took it off confluence and thought it means that kind of sustainability too!

> title: Write sustainable and evolutionary code

Reviewer 2: ```suggestion
title: Write maintainable, reusable and evolutionary code
```

date: git Last Modified
tags:
  - Ways of working
  - Software design
  - Sustainable coding
  - Reusability
---

{% import "macros/tags.njk" as pageTag %}

Last updated: {{ page.date | postDate }}
{{ pageTag.tags(tags)  }}

> When writing code, keep in mind that this should be clean, reusable and descriptive. Reuse existing artefacts before delivering new ones where possible. This can enable code to be easier to understand and cheaper to modify and reuse.

Reviewer 2: Added reference to maintainability
```suggestion
When writing code, keep in mind that this should be clean, reusable and descriptive. Reuse existing artefacts before delivering new ones where possible. This can enable code to be easier to understand, and cheaper to maintain, modify and reuse.
```

---

## Rationale

> Writing code takes time and effort. Code is read more than it is modified, and modified more that it is written. It is worth spending more time and effort to write code that it is easier to understand and to extend with new functionality. 

Reviewer 2: Just a small typo
```suggestion
Writing code takes time and effort. Code is read more than it is modified, and modified more than it is written. It is worth spending more time and effort to write code that it is easier to understand and to extend with new functionality. 
```

Reusing existing code saves considerable development time and effort at the cost of additional complexity. Writing code that is understandable and modular minimises this complexity so that code can be more effectivly reused. Writing code that is easy to modify allows this complexity to only be added when it is useful.

---

## Applications and Implications

> - Keep your code as simple and understandable as possible. Start with only the code needed for the task at hand. Use refactorings where needed to separate concerns, facilitate reuse, and improve readability when the need arises.
> - Have a comprehensive set of regression tests so that you can safely refactor and add new features. If you also practice Test Driven Development you get this for free.
> - Use design patterns when appropriate and avoid using anti-patterns.
> - Keep code loosely coupled and highly reusable which enables future developments to extend the code more easily.
> - Remember code can always be refactored, so keep code simple and refactor for new use cases only when they arise.
> - Prioritise writing readable code. Less code doesn't always mean more efficient/sustainable, is often harder to read and harder to maintain.
> - Use comments sparingly when they add context or meaning that can't be expressed in the code.
> - Use methodologies such as SOLID Principles or Functional Programming where appropriate. Keep in mind methods already used within the project.

Reviewer 2: One grammar edit, and removed periods from ends of lines to be consistent with how we use bullet points on rest of the site
```suggestion
- Keep your code as simple and understandable as possible. Start with only the code needed for the task at hand. Use refactoring where needed to separate concerns, facilitate reuse, and improve readability when the need arises
- Have a comprehensive set of regression tests so that you can safely refactor and add new features. If you also practice Test Driven Development (TDD) you get this for free
- Use design patterns when appropriate and avoid using anti-patterns
- Keep code loosely coupled and highly reusable which enables future developments to extend the code more easily
- Remember code can always be refactored, so keep code simple and refactor for new use cases only when they arise
- Prioritise writing readable code. Less code doesn't always mean more efficient/sustainable, is often harder to read and harder to maintain
- Use comments sparingly when they add context or meaning that can't be expressed in the code
- Use methodologies such as SOLID Principles or Functional Programming where appropriate. Keep in mind methods already used within the project
```

---
layout: sub-navigation
order: 1
title: Write sustainable and evolutionary code
date: git Last Modified
tags:
  - Ways of working
  - Software design

>   - Sustainable Coding

Author: ```suggestion
  - Sustainable coding
```

  - Reusability
---

{% import "macros/tags.njk" as pageTag %}

Last updated: {{ page.date | postDate }}
{{ pageTag.tags(tags)  }}

When writing code, keep in mind that this should be clean, reusable and descriptive. Reuse existing artefacts before delivering new ones where possible. This can enable code to be easier to understand and cheaper to modify and reuse.

---

## Rationale

> This aides in allowing your code to be further developed and evolved in the future, and reduces development time where code is reused.

Author: ```suggestion
This aides in allowing your code to be further developed and evolved in the future, and reduces development time and costs where code is reused. 
```

---

## Applications and Implications

- Keep it simple (especially at first)

> - Avoid anti-patterns

Author: ```suggestion
```

> - Reduce cost and time to market

Author: ```suggestion
```

> - Promote sustainability

Author: ```suggestion
```

> - Enrichen the community

Author: ```suggestion
```

> - Should use tests to verify your refactorings doesn't affect any existing functionality. A good method of doing this is using Test Driven Development.

Author: ```suggestion
- Have a comprehensive set of regression tests so that you can safely refactor and add new features. If you also practice Test Driven Development you get this for free.
```

> - Use refactoring patterns to make functions repeatable and reliable.

Author: ```suggestion
```

> - Use design patterns when appropriate. Such as SOLID Principles or Functional Programming.

Author: ```suggestion
- Use design patterns when appropriate and avoid using anti-patterns.
```

- Keep code loosely coupled and highly reusable which enables future developments to extend the code more easily.
- Remember code can always be refactored, so keep code simple and refactor for new use cases only when they arise.

> - Less code doesn't always mean more efficient/sustainable, is often harder to read and harder to maintain e.g LINQ in C#.

Author: ```suggestion
- Prioritise writing readable code. Less code doesn't always mean more efficient/sustainable, is often harder to read and harder to maintain.
```

- Use comments sparingly when they add context or meaning that can't be expressed in the code.

> 

Author: ```suggestion
- Use methodologies such as SOLID Principles or Functional Programming where appropriate. Keep in mind methods already used within the project.
```
